tests/form_params_test.go

```diff
@@ -0,0 +1,365 @@
+				"\r\n" +
+				"veggie=potato\r\n" +
+				"\r\n"),
+			want: want{
+				contentLen:    "13",
+				queryParamVal: "potato",
+			},
+		},
+		{
+			name: "List of same value in Content-Length Header",
+			req: []byte("POST / HTTP/1.1\r\n" +
+				"Host: localhost:8080\r\n" +
+				"Content-Type: application/x-www-form-urlencoded; charset=utf-8\r\n" +
+				"Content-Length: 13, 13\r\n" +
+				"\r\n" +
+				"veggie=potato\r\n" +
+				"\r\n"),
+			want: want{
+				contentLen:    "13",
+				queryParamVal: "potato",
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), test.req, func(req *http.Request) {
+				if contentLen := req.Header["Content-Length"][0]; contentLen != test.want.contentLen {
+					t.Errorf("Expected Content-Length to be 13 but got %s", contentLen)
+				}
+				if val := req.FormValue("veggie"); val != test.want.queryParamVal {
+					t.Errorf("req.FormValue: want potato, got %s", val)
+				}
+			})
+			if err != nil {
+				t.Errorf("MakeRequest(): got err %v, want nil", err)
+			}
+			if !isOKReq(resp) {
+				t.Errorf("response: want %s, got %s", statusOKReq, resp)
+			}
+		})
+	}
+
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: 13\r\n" +
+		"Content-Length: 12\r\n" +
+		"\r\n" +
+		"veggie=potato\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if contentLen := req.Header["Content-Length"][0]; contentLen != "13" {
+			t.Error("MakeRequest(): request should have been rejected when list of values is given to Content-Length, rather inconsistent if you ask me.")
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, resp)
+	}
+}
+
+// Test whether form parameters with Content-Type:
+// application/x-www-form-urlencoded that break percent-encoding will be
+// ignored
+func TestFormParametersBreakUrlEncoding(t *testing.T) {
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded\r\n" +
+		"Content-Length: 10\r\n" +
+		"\r\n" +
+		"veggie=%sc\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if val := req.Form["veggie"]; len(req.Form["veggie"]) != 0 {
+			t.Errorf("req.Form[\"veggie\"]: got %s, want %s.", val, "")
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOKReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOKReq, resp)
+	}
+}
+
+func TestBasicMultipartForm(t *testing.T) {
+	reqBody := "--123\r\n" +
+		"Content-Disposition: form-data; name=\"foo\"\r\n" +
+		"\r\n" +
+		"bar\r\n" +
+		"--123--\r\n"
+	reqBodyLen := len(reqBody)
```

Comment:
Just use `len(reqBody)`, it's two characters more. :)

Reply:
resolved [3786868](url)